packages/next-swc/crates/core/src/relay.rs

```diff
@@ -0,0 +1,297 @@
+use relay_compiler::{create_path_for_artifact, FileCategorizer, FileGroup, ProjectConfig};
+use serde::Deserialize;
+use swc_atoms::JsWord;
+use swc_common::errors::HANDLER;
+use swc_common::FileName;
+use swc_ecmascript::ast::*;
+use swc_ecmascript::utils::{quote_ident, ExprFactory};
+use swc_ecmascript::visit::{Fold, FoldWith};
+
+#[derive(Copy, Clone, Debug, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum RelayLanguageConfig {
+    Typescript,
+    Flow,
+}
+
+struct Relay {
+    file_name: FileName,
+    relay_config_for_tests: Option<ProjectConfig>,
+}
+
+fn pull_first_operation_name_from_tpl(tpl: &TaggedTpl) -> Option<&str> {
+    tpl.tpl
+        .quasis
+        .iter()
+        .filter_map(|quasis| {
+            let split_content = quasis.raw.value.split(" ").collect::<Vec<&str>>();
+
+            let operation = split_content
+                .chunks(2)
+                .filter_map(|slice| {
+                    if slice.len() == 1 {
+                        return None;
+                    }
+
+                    let word = slice[0];
+                    let next_word = slice[1];
+
+                    if word == "query"
+                        || word == "subscription"
+                        || word == "mutation"
+                        || word == "fragment"
+                    {
+                        return Some(next_word);
+                    }
+
+                    None
+                })
+                .next();
+
+            return operation;
+        })
+        .next()
+}
+
+fn build_require_expr_from_path(path: String) -> Expr {
+    Expr::Call(CallExpr {
+        span: Default::default(),
+        callee: quote_ident!("require").as_callee(),
+        args: vec![Lit::Str(Str {
+            span: Default::default(),
+            value: JsWord::from(path),
+            has_escape: false,
+            kind: Default::default(),
+        })
+        .as_arg()],
+        type_args: None,
+    })
+}
+
+impl Fold for Relay {
+    fn fold_expr(&mut self, expr: Expr) -> Expr {
+        let expr = expr.fold_children_with(self);
+
+        match &expr {
+            Expr::TaggedTpl(tpl) => {
+                if let Some(built_expr) = self.build_call_expr_from_tpl(tpl) {
+                    built_expr
+                } else {
+                    expr
+                }
+            }
+            _ => expr,
+        }
+    }
+}
+
+#[derive(Debug)]
+enum BuildRequirePathError<'a> {
+    FileNameNotReal,
+    MultipleSourceSetsFound {
+        source_set_names: Vec<SourceSetName>,
+        path: &'a PathBuf,
+    },
+    ProjectNotFoundForSourceSet {
+        source_set_name: SourceSetName,
+    },
+    FileNotASourceFile,
+    CouldNotCategorize {
+        err: Cow<'static, str>,
+        path: String,
+    },
+}
+
+// This is copied from https://github.com/facebook/relay/blob/main/compiler/crates/relay-compiler/src/build_project/generate_artifacts.rs#L251
```

Comment:
This will require the relevant license to be mentioned as an inline comment. Relevant file: https://github.com/facebook/relay/blob/main/LICENSE

Reply:
@timneutkens Is the convention to just mention that it's behind an MIT license or should I copy the whole license? Is there an example elsewhere in the codebase?

Reply:
We usually copy the whole license. Here's an example: https://github.com/vercel/next.js/blob/canary/packages/next-swc/crates/napi/src/lib.rs#L1-L27